third_party/move/move-bytecode-verifier/src/signature_v2.rs

```diff
@@ -0,0 +1,984 @@
+                            return map_err(Err(PartialVMError::new(
+                                StatusCode::INVALID_SIGNATURE_TOKEN,
+                            )
+                            .with_message("reference not allowed".to_string())));
+                        }
+                        map_err(self.verify_signature_in_context(&ability_context, *idx))?;
+
+                        entry.insert(());
+                    }
+                },
+
+                // List out the other options explicitly so there's a compile error if a new
+                // bytecode gets added.
+                Pop | Ret | Branch(_) | BrTrue(_) | BrFalse(_) | LdU8(_) | LdU16(_) | LdU32(_)
+                | LdU64(_) | LdU128(_) | LdU256(_) | LdConst(_) | CastU8 | CastU16 | CastU32
+                | CastU64 | CastU128 | CastU256 | LdTrue | LdFalse | Call(_) | Pack(_)
+                | Unpack(_) | ReadRef | WriteRef | FreezeRef | Add | Sub | Mul | Mod | Div
+                | BitOr | BitAnd | Xor | Shl | Shr | Or | And | Not | Eq | Neq | Lt | Gt | Le
+                | Ge | CopyLoc(_) | MoveLoc(_) | StLoc(_) | MutBorrowLoc(_) | ImmBorrowLoc(_)
+                | MutBorrowField(_) | ImmBorrowField(_) | MutBorrowGlobal(_)
+                | ImmBorrowGlobal(_) | Exists(_) | MoveTo(_) | MoveFrom(_) | Abort | Nop => (),
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Checks if a function definition is well-formed.
+    fn verify_function_def(&self, fdef: &FunctionDefinition) -> PartialVMResult<()> {
+        let code = match &fdef.code {
+            Some(code) => code,
+            None => return Ok(()), // No need to check native functions
+        };
+
+        let fh = self.resolver.function_handle_at(fdef.function);
+
+        self.verify_code(&fh.type_parameters, code)
+    }
+
+    /// Checks if all function defs are well-formed.
+    fn verify_function_defs(&self) -> PartialVMResult<()> {
+        for (idx, fdef) in self.resolver.function_defs().unwrap().iter().enumerate() {
+            self.verify_function_def(fdef)
+                .map_err(|err| err.at_index(IndexKind::FunctionDefinition, idx as u16))?;
+        }
+        Ok(())
+    }
+
+    /// Checks if a struct definition is well-formed.
+    ///
+    /// A struct definition is well-formed if
+    /// - All field types are well-formed within the struct context.
+    /// - No phantom type parameters appear in non-phantom positions.
+    ///
+    /// Time complexity: `O(total_size_of_field_types)`
+    fn verify_struct_def(&self, struct_def: &StructDefinition) -> PartialVMResult<()> {
+        let fields = match &struct_def.field_information {
+            StructFieldInformation::Native => return Ok(()),
+            StructFieldInformation::Declared(fields) => fields,
+        };
+        let struct_handle = self.resolver.struct_handle_at(struct_def.struct_handle);
+        let context: Vec<_> = struct_handle.type_param_constraints().collect();
```

Comment:
nit: I don't think we need collect here?

Reply:
Good catch. I'll convert it directly into a bitset.